tests/volume_ops/volume_ops_test.go

```diff
@@ -0,0 +1,158 @@
+		var err error
+		var contexts []*scheduler.Context
+		expReplMap := make(map[*volume.Volume]int64)
+		for i := 0; i < Inst().ScaleFactor; i++ {
+			contexts = append(contexts, ScheduleAndValidate(fmt.Sprintf("volrepldown-%d", i))...)
+		}
+
+		Step("get volumes for all apps in test and decrease replication factor", func() {
+			for _, ctx := range contexts {
+				var appVolumes []*volume.Volume
+				Step(fmt.Sprintf("get volumes for %s app", ctx.App.Key), func() {
+					appVolumes, err = Inst().S.GetVolumes(ctx)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(appVolumes).NotTo(BeEmpty())
+				})
+
+				Step(
+					fmt.Sprintf("repl decrease volume driver %s on app %s's volumes: %v",
+						Inst().V.String(), ctx.App.Key, appVolumes),
+					func() {
+						for _, v := range appVolumes {
+							errExpected := false
+							currRep, err := Inst().V.GetReplicationFactor(v)
+							Expect(err).NotTo(HaveOccurred())
+							if currRep == 1 {
+								errExpected = true
+							}
+							expReplMap[v] = int64(math.Max(1, float64(currRep)-1))
+							err = Inst().V.SetReplicationFactor(v, currRep-1)
+							if !errExpected {
+								Expect(err).NotTo(HaveOccurred())
+							}
+
+						}
+					})
+				Step(
+					fmt.Sprintf("validate successful repl decrease"),
+					func() {
+						time.Sleep(1 * time.Minute)
+						for _, v := range appVolumes {
+							newRepl, err := Inst().V.GetReplicationFactor(v)
+							Expect(err).NotTo(HaveOccurred())
+							Expect(newRepl).To(Equal(expReplMap[v]))
+							Inst().V.ValidateVolumeSetup(v)
+						}
+					})
+			}
+		})
+
+		Step("destroy apps", func() {
+			opts := make(map[string]bool)
+			opts[scheduler.OptionsWaitForResourceLeakCleanup] = true
+			for _, ctx := range contexts {
+				TearDownContext(ctx, opts)
+			}
+		})
+
+	})
+})
+
+// Volume replication increase
+var _ = Describe("VolumeReplicationIncrease", func() {
+	It("has to schedule apps and increase replication factor on all volumes of the apps", func() {
+		var err error
+		var contexts []*scheduler.Context
+		expReplMap := make(map[*volume.Volume]int64)
+		for i := 0; i < Inst().ScaleFactor; i++ {
+			contexts = append(contexts, ScheduleAndValidate(fmt.Sprintf("volreplup-%d", i))...)
+		}
+
+		Step("get volumes for all apps in test and increase replication factor", func() {
+			for _, ctx := range contexts {
+				var appVolumes []*volume.Volume
+				Step(fmt.Sprintf("get volumes for %s app", ctx.App.Key), func() {
+					appVolumes, err = Inst().S.GetVolumes(ctx)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(appVolumes).NotTo(BeEmpty())
+				})
+
+				Step(
+					fmt.Sprintf("repl increase volume driver %s on app %s's volumes: %v",
+						Inst().V.String(), ctx.App.Key, appVolumes),
+					func() {
+						for _, v := range appVolumes {
+							errExpected := false
+							currRep, err := Inst().V.GetReplicationFactor(v)
+							Expect(err).NotTo(HaveOccurred())
+							if currRep == 3 {
```

Comment:
This check for expecting error for repl more than 3 is very px specific.. not all volumes drives have a max repl factor of 3

Reply:
Not sure that is true. I think 3 is more a standard than exception. Even otherwise, elsewhere we have fine tuned APIs based on PX behavior like the return types of HaLevel. How is this different?

Reply:
If anything, we may create a map of drivertype --> constants like this one. For PX, the max repl = 3 and min repl = 1. For XYZDriver it may be whatever. We don't have XYZDriver yet but defining the map/constants may make it easy to add different types later.

Reply:
HaLevel/repl factor is still an integer.. anyone can convert to int64 as it is not a custom type.
For now, I think keep the constant as is instead of making it complex. We or someone else will do it when needed.

Reply:
Create a volume driver interface `GetMaxSupportedReplicationFactor()` and have that return 3 for PX.